fdbserver/DataDistributionQueue.actor.cpp

```diff
@@ -1424,6 +1494,108 @@ ACTOR Future<Void> dataDistributionRelocator(DDQueueData* self, RelocateData rd,
 	}
 }
 
+inline double getWorstCpu(const HealthMetrics& metrics, const std::vector<UID>& ids) {
+	double cpu = 0;
+	for (auto& id : ids) {
+		if (metrics.storageStats.count(id)) {
+			cpu = std::max(cpu, metrics.storageStats.at(id).cpuUsage);
+		} else {
+			// assume the server is too busy to report its stats
+			cpu = std::max(cpu, 100.0);
+			break;
+		}
+	}
+	return cpu;
+}
+
+// Move the shard with the top K highest read density of sourceTeam's to destTeam if sourceTeam has much more read load
+// than destTeam
+ACTOR Future<bool> rebalanceReadLoad(DDQueueData* self,
+                                     int priority,
+                                     Reference<IDataDistributionTeam> sourceTeam,
+                                     Reference<IDataDistributionTeam> destTeam,
+                                     bool primary,
+                                     TraceEvent* traceEvent) {
+	if (g_network->isSimulated() && g_simulator.speedUpSimulation) {
+		traceEvent->detail("CancelingDueToSimulationSpeedup", true);
+		return false;
+	}
+
+	state std::vector<KeyRange> shards = self->shardsAffectedByTeamFailure->getShardsFor(
+	    ShardsAffectedByTeamFailure::Team(sourceTeam->getServerIDs(), primary));
+	traceEvent->detail("ShardsInSource", shards.size());
+	// For read rebalance if there is just 1 hot shard remained, move this shard to another server won't solve the
+	// problem.
+	// TODO: This situation should be solved by split and merge
+	if (shards.size() <= 1) {
+		traceEvent->detail("SkipReason", "NoShardOnSource");
+		return false;
+	}
+
+	// check lastAsSource, at most 10% of shards can be moved within a sample period
```

Comment:
It would be useful for future readers to have the rationale for this described in a comment.